### PR TITLE
Spec Proposal to standardize HOCON path based configuration pattern

### PR DESCRIPTION
**Pull request by Arkatufus**
This will be used to standardize how Akka.Hosting addresses a very common HOCON configuration pattern.
This allows for a HOCON-less programmatical setup replacement for the HOCON path used to configure the HOCON property. 

The pattern:
```text
# This HOCON property references to a config block below
akka.discovery.method = akka.discovery.config

akka.discovery.config {
    class = "Akka.Discovery.Config.ConfigServiceDiscovery, Akka.Discovery"
    # other options goes here
}
```

Example implementation:
```csharp
// The base class for the option
public abstract class DiscoveryOptionBase : IOption
{ }

// The actual option implementation
public class ConfigOption : DiscoveryOptionBase
{
    // Actual option implementation here
    public Type Class => typeof(ConfigServiceDiscovery);

    public string ConfigPath => "akka.discovery.config";

    public void Apply(AkkaConfigurationBuilder builder, Setup setup = null)
    {
        // Modifies Akka.NET configuration either via HOCON or setup class
        builder.AddHocon($"akka.discovery.method = {ConfigPath}", HoconAddMode.Prepend);

        // Rest of configuration goes here
    }
}

 // Akka.Hosting extension implementation
 public static AkkaConfigurationBuilder WithDiscovery(
     this AkkaConfigurationBuilder builder,
     DiscoveryOptionBase discOption)
 {
     var setup = new DiscoverySetup();

     // gets called here
     discOption.Apply(builder, setup);
 }        
```

### CODE REVIEW

**Comment by Aaronontheweb** [comment]
```csharp
 // The base class for the option
 public abstract class DiscoveryOptionBase : IOption
 { }

 // The actual option implementation
 public class ConfigOption : DiscoveryOptionBase
 {
     // Actual option implementation here
     public void Apply(AkkaConfigurationBuilder builder)
     {
         // Modifies Akka.NET configuration either via HOCON or setup class
         builder.AddHocon($"akka.discovery.method = {ConfigPath}", HoconAddMode.Prepend);

         // Rest of configuration goes here
     }
 }

 // Akka.Hosting extension implementation
 public static AkkaConfigurationBuilder WithDiscovery(
     this AkkaConfigurationBuilder builder,
     DiscoveryOptionBase discOption)
 {
     var setup = new DiscoverySetup();

     // gets called here
     discOption.Apply(builder, setup);
 }        
```

I'm not quite seeing the big picture here - how is this an improvement over just prepending config to the front of the HOCON fallback chain? Is the idea here that we instantiate a class that populates the values instead of reading them from a file?

**Comment by Arkatufus** [comment]
Yes, this is to avoid having to provide a HOCON path inside `Akka.Hosting`.

In the example, instead of having the user to remember to use `akka.discovery.config` as the valid string value for the method, they provide the `IOption` implementation instance instead, it is the `IOption` implementation responsibility to provide the HOCON path value value

**Comment by Arkatufus** [comment]
Updated the sample code in the descripition.
Yes, this API is ment for the developers, not the user. And yes, this is just moving the problem from the user to the developer.
The goal is to make it easier for the user to configure Akka.NET via Akka.Hosting, the developer will still need an intimate knowledge of how things are being wired under the hood.

**Comment by Arkatufus** [comment]
It also acts as a "scaffold" for the developer. as in, "if you encounter this kind of HOCON pattern when you create your extension method, this is our standardized way of handling it" so they don't have to re-invent the wheel every time

**Comment by Arkatufus** [comment]
The double abstraction is also intentional, the `IOption` class is a "helper" interface for the developer to use to build their base option class.
The base option class is what the user see, it is the base type for all the options the user can use to configure the extension method and/or `Setup` class

File: src/Akka.Hosting/IOption.cs [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+    ///         var setup = new DiscoverySetup();
+    /// 
+    ///         // gets called here
+    ///         discOption.Apply(builder, setup);
+    ///     }
+    /// </code>
+    /// </example>
+    public interface IOption
+    {
+        public interface IOption
```
Comment (Aaronontheweb) [inline src/Akka.Hosting/IOption.cs:65]: Did you mean to nest interfaces here?

File: src/Akka.Hosting/IOption.cs [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+    ///         DiscoveryOptionBase discOption)
+    ///     {
+    ///         var setup = new DiscoverySetup();
+    /// 
+    ///         // gets called here
+    ///         discOption.Apply(builder, setup);
+    ///     }
+    /// </code>
+    /// </example>
+    public interface IOption
```
Comment (Aaronontheweb) [inline src/Akka.Hosting/IOption.cs:63]: I'd rename this to `IHoconOption` so we don't clash with Micrsoft.Extensions.Configuration

**Review by Aaronontheweb — COMMENTED** [review]
Left some comments

File: src/Akka.Hosting/IOption.cs (hunk 1 above)
Comment (Arkatufus) [reply src/Akka.Hosting/IOption.cs:65]: DOH! I will fix that, thanks

**Comment by Aaronontheweb** [comment]
You know what would really help me here? Refactoring one of our existing configurations to use this - that would help me see, end-to-end, what's going on and what the costs / benefits are.

**Comment by Arkatufus** [comment]
Alright, I'll try to do that.

File: src/Akka.Cluster.Hosting/SBR/SplitBrainResolverOption.cs [hunk 3]
```diff
@@ -84,10 +169,13 @@ public sealed class KeepOldestOption : SplitBrainResolverOption
...
     {
+        public override string ConfigPath => SplitBrainResolverSettings.LeaseMajorityName;
+
         /// <summary>
-        /// The absolute HOCON path to the <see cref="Lease"/> implementation HOCON configuration block
+        /// An class instance that extends <see cref="LeaseOptionBase"/>, used to configure the lease provider used in this
+        /// <see cref="LeaseMajority"/> strategy.
         /// </summary>
-        public string LeaseImplementation { get; set; }
+        public LeaseOptionBase LeaseImplementation { get; set; }
```
Comment (Arkatufus) [inline src/Akka.Cluster.Hosting/SBR/SplitBrainResolverOption.cs:178]: This is an example of direct use of this standard.
Before, the user need to remember that the lease implementation for `Akka.Coordination.KubernetesApi` is `"akka.coordination.lease.kubernetes"`, now they can actually use intellisense and find `KubernetesLeaseOption` (fake name, not implemented yet) instead.

**Review by Aaronontheweb — APPROVED** [review]
LGTM

File: src/Akka.Cluster.Hosting/SBR/SplitBrainResolverOption.cs [hunk 4]
```diff
@@ -38,10 +46,37 @@ public abstract class SplitBrainResolverOption
     /// </summary>
     public sealed class StaticQuorumOption : SplitBrainResolverOption
     {
+        public override string ConfigPath => SplitBrainResolverSettings.StaticQuorumName;
```
[the remark this answers is not in the document]
Comment (Arkatufus) [reply src/Akka.Cluster.Hosting/SBR/SplitBrainResolverOption.cs:49]: Whatever the plugin/module requires, it is the Akka.Hosting module developer duty to know the required HOCON path type to return here.
Historically, we do not have any standard regarding this value, it is absolute in some case, but relative in other.

**Pull request merged**